hw/ip/aes/pre_sca/alma_post_syn/verify_aes_post_syn.sh

```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# Script to formally verify the masking of the post-synthesis AES S-box netlist with CocoAlma.
+
+set -e
+
+# If the output dir exists, remove files from previous run. Otherwise, create the folder.
+if [ -d ${OUTPUT_DIR} ]; then
+  rm ${OUTPUT_DIR}/circuit*
+  rm ${OUTPUT_DIR}/labels.txt
+  rm ${OUTPUT_DIR}/tmp.vcd
+else
+  echo "Creating ${OUTPUT_DIR}"
+  mkdir -p ${OUTPUT_DIR}
+fi
+
+# Create an ad hoc techlib .v file by patching complex cells from given patch file.
+# The following scripts reads the required filenames from environment variables, therefore we
+# are not passing them as arguments.
+echo "Patching up library cells"
+${REPO_TOP}/hw/ip/aes/pre_sca/alma_post_syn/simplify_techlib.py 
+
+# `TOP_MDOULE` and `TESTBENCH` parameters are loaded as environment variables.
+echo "Verifying post-synthesis netlist ${TOP_MODULE} using CocoAlma"
+
+# Parse
+${ALMA_TOP}/parse.py --top-module ${TOP_MODULE} \
+--log-yosys \
+--synthesis-template ${REPO_TOP}/hw/ip/aes/pre_sca/alma_post_syn/yosys_synth_template.txt \
+--netlist ${OUTPUT_DIR}/circuit.v \
+--label ${OUTPUT_DIR}/labels.txt \
+--json ${OUTPUT_DIR}/circuit.json \
+--source ${AES_SBOX_NETLIST_FILE} ${OUTPUT_DIR}/modified_cell_library.v
+
+# Label
+sed -i 's/\(data_i\s\[\)\([0-9]\+:0\)\(\]\s=\s\)unimportant/\1\2\3secret \2/g' ${OUTPUT_DIR}/labels.txt
+sed -i 's/\(mask_i\s\[\)\([0-9]\+:0\)\(\]\s=\s\)unimportant/\1\2\3secret \2/g' ${OUTPUT_DIR}/labels.txt
+sed -i 's/\(prd_i\s\[[0-9]\+:0\]\s=\s\)unimportant/\1static_random/g' ${OUTPUT_DIR}/labels.txt
```

Comment:
IIRC when using `static_random` you assume that the `prd_i` input remains constant throughout the execution. This is fine for a single S-Box evaluation. And it's actually more accurate for how we use the S-Box (we're updating the `prd_i` inputs once per execution). But I would expect this to fail when running the verification for double the cycles and changing the testbench to do two evaluations. Out of curiosity, did you ever try this?

Reply:
I have not tried this, because running two S-box executions with the same randomness seemed difficult with CocoAlma. Or at least I could not see a quick solution to run this in my mind.

What I did instead was, marking one of the random bits as unimportant. Specifically, I tried the following in labels.txt:
```
prd_i [26:0] = static_random
prd_i [27] = unimportant
```
which ends up with a failure:
```
The execution is not secure, here are some leaks:
leak 0: (cycle: 5, cell: xor out_mask_basis_x[3], id: 2562) 
5 stable  xor out_mask_basis_x[3] vars   : ['s0:0', 's0:1']
5 stable  xor out_mask_basis_x[3] signals: data_i[0] ^ mask_i[0]
5 trans   xor out_mask_basis_x[3] vars   : ['s0:0', 's0:1']
5 trans   xor out_mask_basis_x[3] signals: data_i[0] ^ mask_i[0]
```

Do you think this is sufficient?

Reply:
In general, this seems like an interesting scenario and after discussion offline we concluded the tool would probably not be able to detect if two evaluations would use the same randomness. I am resolving this discussion.

The example you did is also relevant. One of the randomness inputs used in the last S-Box pipeline stage is marked as non-random. Indeed, the tool detects a leak in the last cycle where an intermediate result depending on bit 27 is recombined with something else. So everything is as suspected. When experimenting with the tool back when I was implementing the various AES S-Boxes, I always kept doing such little test to convince myself both the tool and I had the same understanding of how things should be :-)